### ISSUE TITLE
Add a `manual-if-remote` tag

### ISSUE THREAD

**Issue by shs96c** [bazelbuild/bazel#18605]
### Description of the feature request:

It would be helpful to be able to tag targets with `manual-if-remote` to exclude them remote execution and evaluation with wildcard builds

### What underlying problem are you trying to solve with this feature?

Remote builds are often done between mixed host and remote OSs (eg. a macOS developer machine connecting to a BuildBarn or EngFlow cluster running Linux) Getting a source tree to build both locally and remotely is a challenging prospect.

The `local` tag (and similar tags, such as `no-remote`) will schedule work on the local machine, but OS and CPU mismatches frequently cause those targets to fail. Assigning a tag based on a `select` isn't possible because `tags` are nonconfigurable.

As an example of where this is useful, Selenium tests against a range of browsers. Some of these (chrome and firefox) we can "pin" as dependencies and use remotely, but some (such as webkitdriver) can only be run locally if the developer has set up their machine properly. This means that we'd like to be able to schedule the tests when running locally, but to skip them entirely when running remotely.

We _could_ modify our test runner to check for the presence of the binaries and use something like an `Assume` to skip the tests once they start executing, but that means scheduling work we know won't be necessary and slows down our builds. It would be better to skip them entirely.

It should go without saying that this specific example isn't the only use-case for a `manual-if-remote` tag :) 

### Which operating system are you running Bazel on?

macOS and Linux

### What is the output of `bazel info release`?

6.2.1

**Comment by meisterT** [comment]
cc @lberki for his thoughts

**Comment by katre** [comment]
This is pretty complex logic for a tag.

Ideally, as @lberki says, we'd have integrated execution strategy and platforms enough that Bazel could tell that target X can't be executed with the current execution strategy, so it'd be reported as skipped. Since we're not there, we need a workaround. Is `manual-if-remote` the workaround that will be easiest to deprecate and remove when we no longer need it?

Would something with execution properties be cleaner?
```
cc_test(
    name = "foo_test",
    exec_properties = {
        "remote": "manual",
    },
)
```

This is a bit different than what we use these for now, whereas tags absolutely are used for this now, but at least it's more structured?

**Comment by shs96c** [comment]
I think it's a little more subtle than just using platforms. When running on CI, it's possible that the local and remote machines are running the same platform, but we'd not expect tests (for example) to pass remotely.

**Comment by lberki** [comment]
What is the difference between the local and remote end that would make the test fail on one and pass on the other?

I, and I believe @katre are both using the term "platform" in its Bazel sense; i.e. not just CPU/OS/etc/., but something that captures every property of the system that might affect build/test outcomes (including browser versions and the like)

**Comment by shs96c** [comment]
The differences I've seen are things like:

* Local machines have a Docker socket, remote ones do not
* Remote machines have stricter network access permissions, making external sites unavailable
* Remote workers are stripped down to the bare minimum, and expect everything to be in the CAS. Local machines have locally installed software
* Relatedly, starting a local server for some part of the system (eg. a database) and running tests against that locally is common pattern. That doesn't work with remote builds, as the software is missing.

If the expected approach is for users to use platforms to handle this, the documentation needs to be clear on how to do this, particularly when local and remote systems share the same OS, cpu, and so on.

**Comment by lberki** [comment]
Oh, I see. I suspected that the meaningful difference was that the two machines have different versions of browsers / Selenium / etc., but what's on your list is probably indeed better not captured in the Bazel platform.

I'll defer to @katre then.

**Comment by katre** [comment]
I agree that the use cases you list are a little different that how we typically use platforms.

- Some systems have running docker daemons: maybe this is a constraint? Maybe it's a build setting?
- Some systems don't have external network: this definitely feels like a build setting
- Locally installed software, and locally running test systems feel like non-hermeticity to me. I'm not sure how to clearly express "on some platforms non-hermetic builds will work".

I don't have any good answers. The proposed "manual-if-remote" tag feels like a lot of magic processing in the already magical tag system.

**Comment by meisterT** [comment]
(Sorry John for misclassifying this issue.)

@shs96c there seems to be a lot of complexity here. In addition to what @katre brought up, one of the issues that we see is that target selection happens early on in the build while strategy selection happens during action execution. We don't see a way to implement this without adding a lot of complexity to the code base. Happy to be convinced otherwise...

**Comment by shs96c** [comment]
Surely this is just the same kind of logic that is already present for `exclusive-if-local`?

**Comment by tjgq** [comment]
`exclusive-if-local` does not affect whether the action will run, only when to schedule it. The proposed `manual-if-remote` tag affects whether the action will run at all. By the time we decide which strategy to use for an action, we're already committed to executing it, so this feature would require the ability to undo the decision. (It's also possible that there's a clever way to decide sufficiently early, but I don't immediately see how.)
